### ISSUE TITLE
Manual GPS location entry to prevent GPS blockage

### ISSUE THREAD

**Issue by E3V3A** [SecUpwN/Android-IMSI-Catcher-Detector#309]
One weakness of our app is that it's downloaded data and map function relies on you having a well defined GPS location,which normally means a working GPS position lock-on. Unfortunately it is very easy to jam GPS signal today with [devices](http://www.pki-electronic.com/products/product-category/jamming-systems/) like this:

![6080-500x500 1](https://cloud.githubusercontent.com/assets/194392/5989888/0451fe9e-a9a0-11e4-917c-48b66971044b.jpg)

Thus we need a way to manually enter the current GPS data into our app. We should be able to do this under settings, for example. Or asked for, when the GPS is non-functioning.

**Comment by SecUpwN** [comment]
>Or asked for, when the GPS is non-functioning.

I'd prefer it that way. A window should appear tellling the user what just happened and ask for the GPS location. But how shall a user know the exact GPS coordinates without enabling GPS? Hard thing to know!

**Comment by E3V3A** [comment]
Not really. Open the map, and drag the guy to your current location. Easy speaking. The practical implementation is the hard part.

**Comment by tobykurien** [comment]
The hard part here is that the event usually happens in the background service, when the app isn't even showing. This means a notification should popup asking the user for GPS coords, but the user might only get to it once they've moved elsewhere. Also, multiple of these events may have occurred in the meantime, with the user unable to remember where they were.

**Comment by E3V3A** [comment]
Sorry @tobykurien, I'm not following you. What do you mean with "event"? Do you mean GPS lock-on? We only need this once, unless we move 10's of Kilometers away, since it's only needed for BTS (DBe_import) download and showing your current position on the map, if you need to look at that. If you don't care about the map, then I don't see a problem. But if you're dependent on the map and relative positions, then we need another way. How about using WiFi localization? Also, relative measurements to the (already downloaded BTS positions) could also be used to give an approximate location, as I believe that huge transparent circle does. (although that circle is using GPS accuracy and not BTS accuracy.) Also, CDMA provide BTS GPS positions, but I don't know if LTE does.

**Comment by E3V3A** [comment]
Upgraded to "bug" status so people can start thinking to implement how to manually select and force current location, so GPS can be disabled. Very useful when one only want to detect silent type-0 SMS and not track cells.

**Comment by marvinmarnold** [comment]
I think it would be interesting to avoid user input as much as possible. I'm new to the code, but based on this assumption:

> it's only needed for BTS (DBe_import) download and showing your current position on the map

For the download issue, could use the last known GPS and download a larger area around that? I think that would be a good approximation if used in conjunction with a potentially known nearby BTS.
WiFi Sounds good too, a quick search turned up https://wigle.net/.

No clue about the BTS GPS in LTE or CDMA.

**Comment by marvinmarnold** [comment]
For WiFi, I think we already do this through https://github.com/SecUpwN/Android-IMSI-Catcher-Detector/blob/8a061dc9f18265ac550b3d80a2232e5fbd5208c7/app/src/main/java/com/SecUpwN/AIMSICD/service/LocationTracker.java#L105

Not sure how different Android's WiFi DB is from something like Wigle...

**Comment by E3V3A** [comment]
@marvinmarnold 

> For the download issue, could use the last known GPS...

Yes

>... and download a larger area around that?

No

> Can you explain the relative importance for type-0 SMS some?

No, it's already been discussed at excruciating lengths elsewhere and on XDA thread. Look there.

Also, if your GPS was being jammed, what make you think that WiFi would not?

**Comment by marvinmarnold** [comment]
> Also, if your GPS was being jammed, what make you think that WiFi would not?
I was also wondering the same thing. Though, if we assume internet and GPS are always down at the same time, then I don't understand the use case this is trying to address. Is it exclusively for when you are in an area you already have downloaded BTS for but no GPS?

**Comment by E3V3A** [comment]
> Is it exclusively for when you are in an area you already have downloaded BTS for but no GPS?

Yes, more or less, and also beacuse GPS lock-on attempts interferes with baseband operations (it is a BB  operation) on some devices, blocking the info sent to AP.

**Comment by E3V3A** [comment]
In addition, many older phones may not have or have a broken GPS.

**Comment by DJaeger** [comment]
At least the app should fall back to "network location" if GPS is (activated, but) not available.
But we can think about what to do if both are not available

**Comment by E3V3A** [comment]
How would you know it's not available when activated?

**Comment by DJaeger** [comment]
By timeout

**Comment by E3V3A** [comment]
timeout of what?

**Comment by DJaeger** [comment]
Sorry.
The app should fall back to network location if the request of GPS/fine location gets a timeout.

**Comment by ga900** [comment]
That noise generators do not have a big range, 5-10 eventually 20m. For 50m radius you will need at least a briefcase sized jammer. GPS is pretty hard to jam and expensive if you try to do it right. There is one trick with coding sequences, but it is expensive to produce and only in that case you can jam big area with small power. Btw LTE has similar vulnerability described in some papers found at some .edu domain. So forget about jamming gps and if you can not get a gps fix just use last known position for database download.

**Comment by E3V3A** [comment]
@ga900  Thanks for your valuable input. You're right, and if jammed properly, you'd be in a place rather f-d anyway. However, manual GPS entry could be useful anyway, for example, for being able to download for a region you have not yet entered, or for indoor operations in a large plaza where femto-cells are covering, but GPS not reachable. Anyway, your arguments is convincing enough for removing bug label.

**Comment by TPS** [comment]
Do you all want to hook a GPS/map intent to make this manual GPS entry simpler?  Or does AICD already do that? https://github.com/guardianproject/LocationPrivacy seems to have secure code that could be co-opted for this.

**Comment by ga900** [comment]
@E3V3A How big "large plaza" should be to have more than one LAC? As far as I remember, Munchen messe is on T-mobile in one LAC. There is possibility that all femto-cells within shopping mall/plaza are in one LAC that is different from btses

**Comment by ga900** [comment]
@E3V3A I agree that manual download data for a new area would be a nice feature, so you can go prepared.Concerning large buildings, how big "large plaza" should be to have more than one LAC? As far as I remember, Munchen messe is on T-mobile in one LAC. There is possibility that all femto-cells within shopping mall/plaza are in one LAC that is different from btses outside, but such solution requires more work to set-up and would generate tremendous amount of unnecesery traffic from frequent location updates if parameters are not set properly (that requires a lots of expirience with similar situations and fine-tuning still takes a few days of work per cell) so it is easier (read cheaper) to put everything in the same lac. So I would say solution in that case is increase detection sensitivity and monitor lac change. Sorry for possible typos, message is written on mobile.

**Comment by E3V3A** [comment]
> Do you all want to hook a GPS/map intent to make this manual GPS entry simpler? 
> Or does AICD already do that?
@TPS: Not sure how that is done in practice... (I doubt we do that.)
